core/src/pow/cuckarood.rs

```diff
@@ -0,0 +1,191 @@
+use crate::pow::error::{Error, ErrorKind};
+use crate::pow::siphash::siphash_block;
+use crate::pow::{PoWContext, Proof};
+use crate::global;
+
+/// Instantiate a new CuckaroodContext as a PowContext. Note that this can't
+/// be moved in the PoWContext trait as this particular trait needs to be
+/// convertible to an object trait.
+pub fn new_cuckarood_ctx<T>(
+	edge_bits: u8,
+	proof_size: usize,
+) -> Result<Box<dyn PoWContext<T>>, Error>
+where
+	T: EdgeType + 'static,
+{
+	let params = CuckooParams::new(edge_bits, proof_size)?;
+	Ok(Box::new(CuckaroodContext { params }))
+}
+
+/// Cuckaroo cycle context. Only includes the verifier for now.
+pub struct CuckaroodContext<T>
+where
+	T: EdgeType,
+{
+	params: CuckooParams<T>,
+}
+
+impl<T> PoWContext<T> for CuckaroodContext<T>
+where
+	T: EdgeType,
+{
+	fn set_header_nonce(
+		&mut self,
+		header: Vec<u8>,
+		nonce: Option<u32>,
+		_solve: bool,
+	) -> Result<(), Error> {
+		self.params.reset_header_nonce(header, nonce)
+	}
+
+	fn find_cycles(&mut self) -> Result<Vec<Proof>, Error> {
+		unimplemented!()
+	}
+
+	fn verify(&self, proof: &Proof) -> Result<(), Error> {
+		if proof.proof_size() != global::proofsize() {
+			return Err(ErrorKind::Verification(
+				"wrong cycle length".to_owned(),))?;
+		}
+		let nonces = &proof.nonces;
+		let mut uvs = vec![0u64; 2 * proof.proof_size()];
+		let mut ndir = vec![0usize; 2];
+		let mut xor0: u64 = 0;
+		let mut xor1: u64 = 0;
+		let nodemask = self.params.edge_mask >> 1;
+
+		for n in 0..proof.proof_size() {
+			let dir = (nonces[n] & 1) as usize;
+			if ndir[dir] >= proof.proof_size() / 2 {
+				return Err(ErrorKind::Verification("edges not balanced".to_owned()))?;
+			}
+			if nonces[n] > to_u64!(self.params.edge_mask) {
+				return Err(ErrorKind::Verification("edge too big".to_owned()))?;
+			}
+			if n > 0 && nonces[n] <= nonces[n - 1] {
+				return Err(ErrorKind::Verification("edges not ascending".to_owned()))?;
+			}
+			let edge = to_edge!(siphash_block(&self.params.siphash_keys, nonces[n], 25));
+			let idx = 4 * ndir[dir] + 2 * dir;
+			uvs[idx  ] = to_u64!( edge        & nodemask);
+			uvs[idx+1] = to_u64!((edge >> 32) & nodemask);
+			xor0 ^= uvs[idx  ];
+			xor1 ^= uvs[idx+1];
+			ndir[dir] += 1;
+		}
+		if xor0 | xor1 != 0 {
+			return Err(ErrorKind::Verification(
+				"endpoints don't match up".to_owned(),
+			))?;
+		}
+		let mut n = 0;
+		let mut i = 0;
+		let mut j;
+		loop {
+			// follow cycle
+			j = i;
+			for k in (((i % 4) ^ 2)..(2 * self.params.proof_size)).step_by(4) {
+				if uvs[k] == uvs[i] { // find reverse edge endpoint identical to one at i
+					if j != i {
+						return Err(ErrorKind::Verification("branch in cycle".to_owned()))?;
+					}
+					j = k;
+				}
+			}
+			if j == i {
+				return Err(ErrorKind::Verification("cycle dead ends".to_owned()))?;
+			}
+			i = j ^ 1;
+			n += 1;
+			if i == 0 {
+				break;
+			}
+		}
+		if n == self.params.proof_size {
+			Ok(())
+		} else {
```

Comment:
Unnecessary else?

Reply:
necessary, unless you add a return before Ok....

Reply:
Indeed.